### ISSUE TITLE
Add ValidationRule to disallow whitespace in the local-part and domain

### ISSUE THREAD

**Issue by PascalSchumacher** [RohanNagar/jmail#53]
Using JMail `1.4.0` this test passes:

```
@Test
void test() {
    assertTrue(JMail.isValid("a@b .com"));
    assertTrue(JMail.isValid("a. b@c.com"));
}
```

Are these valid email addresses?

The first one lead to a `javax.mail.internet.AddressException: Domain contains control or whitespace in string ''a@b .com''` when I try to use it with java mail.

By the way: Thank you very much for providing JMail. It is the best java libary to validate e-mail adresses! 👍

**Comment by RohanNagar** [comment]
Hi @PascalSchumacher! Thank you for the nice words 🙂

According to [RFC 5322](https://datatracker.ietf.org/doc/html/rfc5322) and [RFC 2822](https://datatracker.ietf.org/doc/html/rfc2822) those are valid (although unusual) email addresses.

Essentially, white-space is allowed between dot-separated parts of the local-part and the domain since RFC 822. This has been [marked as obsolete](https://datatracker.ietf.org/doc/html/rfc2822#section-4.4) in RFC 2822, which means that addresses should not be generated with whitespace, but they **should be accepted as valid** addresses by an email server, which is why JMail considers them as valid by default.

[Appendix A.6.3](https://datatracker.ietf.org/doc/html/rfc5322#appendix-A.6.3) shows a valid example with the obsolete whitespace: `jdoe@machine(comment).  example`

All that being said, in practice you may want to consider these addresses as invalid since the syntax is marked as obsolete. One way you could do that is by adding a rule to your validator that checks to make sure both the local-part and the domain do not contain whitespace:

```java
JMail.strictValidator()
  .withRule(email -> !email.localPart().matches("\\S+") && !email.domain().matches("\\S+"))
  .isValid("a@b .com");
```

 I think it would be a great idea for JMail to provide a rule which will do this kind of check so that you can add it like:
 
 ```java
EmailValidator validator = JMail.strictValidator().disallowWhitespace();
 ```
 
 I will keep this issue open as a to-do for that enhancement.

**Comment by PascalSchumacher** [comment]
Thank you very much for the in-depth explanation.

I wasn't aware that whitespace is allowed. I noticed that email addresses like these are either rejected by Java Mail or by Apache James Mail Server.

I will use the rule you suggested and later update the code to use the build-in check when/if you add it to JMail.

Thanks again for the fast and detailed response.

**Comment by RohanNagar** [comment]
Hi @PascalSchumacher this improvement was made and released in version v1.4.1.

You can use it like the following:

```java
EmailValidator validator = JMail.strictValidator().disallowObsoleteWhitespace();

validator.isValid("a@b .com"); // returns false
```
